ingestion/src/metadata/ingestion/source/database/couchbase/metadata.py

```diff
@@ -0,0 +1,127 @@
+
+from metadata.generated.schema.entity.services.connections.database.couchbaseConnection import (
+    CouchbaseConnection,
+)
+from metadata.generated.schema.entity.services.connections.metadata.openMetadataConnection import (
+    OpenMetadataConnection,
+)
+from metadata.generated.schema.metadataIngestion.workflow import (
+    Source as WorkflowSource,
+)
+from metadata.ingestion.api.source import InvalidSourceException
+from metadata.ingestion.source.database.common_nosql_source import (
+    SAMPLE_SIZE,
+    CommonNoSQLSource,
+)
+from metadata.ingestion.source.database.couchbase.queries import (
+    COUCHBASE_GET_DATA,
+    Couchbase_SQL_STATEMENT,
+)
+from metadata.utils.logger import ingestion_logger
+
+logger = ingestion_logger()
+
+
+class CouchbaseSource(CommonNoSQLSource):
+    """
+    Implements the necessary methods to extract
+    Database metadata from Dynamo Source
+    """
+
+    def __init__(self, config: WorkflowSource, metadata_config: OpenMetadataConnection):
+        super().__init__(config, metadata_config)
+        self.couchbase = self.connection_obj
+
+    @classmethod
+    def create(cls, config_dict, metadata_config: OpenMetadataConnection):
+        config: WorkflowSource = WorkflowSource.parse_obj(config_dict)
+        connection: CouchbaseConnection = config.serviceConnection.__root__.config
+        if not isinstance(connection, CouchbaseConnection):
+            raise InvalidSourceException(
+                f"Expected CouchbaseConnection, but got {connection}"
+            )
+        return cls(config, metadata_config)
+
+    def get_database_names(self) -> Iterable[str]:
+        if self.service_connection.bucket:
+            yield self.service_connection.__dict__.get("bucket")
+        else:
+            buckets = self.couchbase.buckets()
+            for bucket_name in buckets.get_all_buckets():
+                yield bucket_name.name
+
+    def get_schema_name_list(self) -> List[str]:
+        """
+        Method to get list of schema names available within NoSQL db
+        need to be overridden by sources
+        """
+        try:
+            database_name = self.context.database.name.__root__
+            bucket = self.couchbase.bucket(database_name)
+            collection_manager = bucket.collections()
+            self.context.scope_list = collection_manager.get_all_scopes()
+            return [scopes.name for scopes in self.context.scope_list]
+        except Exception as exp:
+            logger.debug(f"Failed to list database names: {exp}")
+            logger.debug(traceback.format_exc())
+        return []
+
+    def get_table_name_list(self, schema_name: str) -> List[str]:
+        """
+        Method to get list of table names available within schema db
+        need to be overridden by sources
+        """
+        try:
+            scope_obj = None
+            for scope_obj in self.context.scope_list:
+                if scope_obj.name == schema_name:
+                    break
```

Comment:
why do we need to traverse through the entire schema list for each schema name??